### PR TITLE
set containers DockerId per task

### PR DESCRIPTION
**Pull request by cyastella**
### Summary

expose container dockerId per task 
### Implementation details

update container runtimeId attribute to container and containerStatechange struct. 
set containerDockerId when container is created.
### Testing

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/engine/docker_task_engine.go [hunk 1]
```diff
@@ -942,6 +942,8 @@ func (engine *DockerTaskEngine) createContainer(task *apitask.Task, container *a
 	container.SetLabels(config.Labels)
 	seelog.Infof("Task engine [%s]: created docker container for task: %s -> %s, took %s",
 		task.Arn, container.Name, metadata.DockerID, time.Since(createContainerBegin))
+	container.SetRuntimeId(metadata.DockerID)
```
Comment (suneyz) [inline agent/engine/docker_task_engine.go]: I noticed there is docker id empty check on line 935, do we want to also include this there or we don't care here that docker id is empty?

File: agent/api/container/container_test.go [hunk 2]
```diff
@@ -479,3 +479,10 @@ func TestPerContainerTimeouts(t *testing.T) {
 	assert.Equal(t, container.GetStartTimeout(), expectedTimeout)
 	assert.Equal(t, container.GetStopTimeout(), expectedTimeout)
 }
+
+func TestSetRuntimeIdInContainer(t *testing.T) {
+	container := Container{
+		RuntimeId: "asdfghjkl1234",
+	}
+	assert.Equal(t, "asdfghjkl1234", container.GetRuntimeId())
```
Comment (sparrc) [inline agent/api/container/container_test.go]: currently no test coverage on SetRuntimeId. Add a call to SetRuntimeId() to change the value then validate with another assert.

File: agent/api/container/container.go [hunk 3]
```diff
@@ -96,6 +96,8 @@ type HealthStatus struct {
 type Container struct {
 	// Name is the name of the container specified in the task definition
 	Name string
+	// RuntimeId is the docker id of the container
+	RuntimeId string
```
Comment (sparrc) [inline agent/api/container/container.go]: Can we change this to RuntimeID?

File: agent/engine/docker_task_engine.go (hunk 1 above)
Comment (cyastella) [reply agent/engine/docker_task_engine.go]: When create the container failed, we leave the container's dockerId as null.

File: agent/engine/docker_task_engine.go [hunk 4]
```diff
@@ -942,6 +942,8 @@ func (engine *DockerTaskEngine) createContainer(task *apitask.Task, container *a
 	container.SetLabels(config.Labels)
 	seelog.Infof("Task engine [%s]: created docker container for task: %s -> %s, took %s",
 		task.Arn, container.Name, metadata.DockerID, time.Since(createContainerBegin))
+	container.SetRuntimeID(metadata.DockerID)
+	seelog.Infof("RuntimeId of the container set when create the container %s", container.GetRuntimeID())
```
Comment (sharanyad) [inline agent/engine/docker_task_engine.go]: I don't think this log statement is necessary. this might get noisy.

File: agent/api/statechange.go [hunk 5]
```diff
@@ -138,6 +140,7 @@ func NewContainerStateChangeEvent(task *apitask.Task, cont *apicontainer.Contain
 	event = ContainerStateChange{
 		TaskArn:       task.Arn,
 		ContainerName: cont.Name,
+		RuntimeID:     cont.RuntimeID,
```
Comment (sharanyad) [inline agent/api/statechange.go]: can you use `cont.GetRuntimeID()` here instead?

Comment (cyastella) [reply agent/api/statechange.go]: yes.

File: agent/engine/docker_task_engine.go [hunk 6]
```diff
@@ -942,6 +942,7 @@ func (engine *DockerTaskEngine) createContainer(task *apitask.Task, container *a
 	container.SetLabels(config.Labels)
 	seelog.Infof("Task engine [%s]: created docker container for task: %s -> %s, took %s",
 		task.Arn, container.Name, metadata.DockerID, time.Since(createContainerBegin))
+	container.SetRuntimeID(metadata.DockerID)
```
Comment (sharanyad) [inline agent/engine/docker_task_engine.go]: this field should probably be part of the [container struct](https://github.com/aws/amazon-ecs-agent/blob/f95f731bce74e6276f9fd8bfbb87a62d3a790ce8/agent/api/container/container.go#L96) and stored in state file to handle agent restarts.

File: agent/api/container/container.go [hunk 7]
```diff
@@ -96,6 +96,8 @@ type HealthStatus struct {
 type Container struct {
 	// Name is the name of the container specified in the task definition
 	Name string
+	// RuntimeID is the docker id of the container
+	RuntimeID string
```
Comment (sharanyad) [inline agent/api/container/container.go]: this probably needs [statefile version](https://github.com/aws/amazon-ecs-agent/blob/fa463fee4c79352636ad2e3851144a9ac0804be2/agent/statemanager/state_manager.go#L91) bump and test added [here](https://github.com/aws/amazon-ecs-agent/blob/dev/agent/statemanager/state_manager_test.go).

File: agent/statemanager/testdata/v23/perContainerRuntimeID/ecs_agent_data.json [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+          "PulledAt": "2018-10-04T18:05:48.445644088Z",
+          "LastUsedAt": "2018-10-04T18:05:48.445645342Z",
+          "PullSucceeded": false
+        }
+      ],
+      "ENIAttachments": null,
+      "IPToTask": {}
+    }
+  },
+  "Version": 20
```
Comment (fenxiong) [inline agent/statemanager/testdata/v23/perContainerRuntimeID/ecs_agent_data.json]: this is not using state file version 23.. i think you need to build the agent with your current changes and run a task, then put the state file here

File: agent/api/statechange_test.go [hunk 9]
```diff
@@ -83,3 +84,21 @@ func TestSetTaskTimestamps(t *testing.T) {
 	assert.Equal(t, t2.UTC().String(), change.PullStoppedAt.String())
 	assert.Equal(t, t3.UTC().String(), change.ExecutionStoppedAt.String())
 }
+
+func TestSetContainerRuntimeID(t *testing.T) {
```
Comment (fenxiong) [inline agent/api/statechange_test.go]: this is not testing any method.. you should probably test NewContainerStateChangeEvent and NewTaskStateChangeEvent?

File: agent/api/ecsclient/client.go [hunk 10]
```diff
@@ -391,6 +392,15 @@ func (client *APIECSClient) buildContainerStateChangePayload(change api.Containe
 		ContainerName: aws.String(change.ContainerName),
 	}
 
+	if change.RuntimeID != "" {
+		if len(change.RuntimeID) > ecsMaxRuntimeIDLength {
```
[the remark this answers is not in the document]
Comment (petderek) [reply agent/api/ecsclient/client.go]: If you make it so trimString returns an unmodified string if input < 255, you can clean up the middle code a lot.

```
statechange.RuntimeID = trimString(change.RuntimeID, maxRuntimeIDLength)

statechange.Reason = trimString(change.Reason, maxReasonLength)

File: agent/api/ecsclient/client.go [hunk 11]
```diff
@@ -38,6 +38,7 @@ import (
 
 const (
 	ecsMaxReasonLength    = 255
+	ecsMaxRuntimeIDLength = 255
```
[the remark this answers is not in the document]
Comment (cyastella) [reply agent/api/ecsclient/client.go]: sync offline

**Pull request merged**